[NT] Support NestedTensor in is_concrete_int

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124803

We run into an error when we try to construct a NT in the graph and return it. One of the errors is with symint handling in is_concrete_int: the full error this PR is here: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py.

Differential Revision: [D58315861](https://our.internmc.facebook.com/intern/diff/D58315861)

davidberard98 committed:
[WIP] Support NT constructed in a graph and returned

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] Support NT constructed in a graph and returned"




[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][NT] Support a NJT constructed in a graph and returned from the graph"


We run into an error when we try to construct a NT in the graph and return it. There's two things fixed here:

1. some symint handling utils don't work with NestedInt (this is fixed by the changes in symbolic_shapes.py. error before the fixes: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py)
2. support tensor_unflatten for non-symbolic tensors. In this case, we haven't run mark_dynamic on the values or offsets, so torch.compile is treating them as static shape tensors. Although this is not ideal (we're almost certainly going to have to recompile on the next batch which will probably have different offsets / different total numel), it shouldn't be an error. In this case, this is relevant for some logic in tensor_unflatten: if we're dealing with a functional tensor or fake tensor and we're trying to tensor_unflatten using "tracing tensors" as the offsets/values, then we know we should be able to reuse the "old" NestedInt from the outer size. Previously, the check for "whether this is a tracing tensor" was "whether there are any symbolic sizes"; but in the case where we haven't run `mark_dynamic`, we don't have any symbolic sizes. Instead, just check if we have a FunctionalTensor or FakeTensor as an indicator of whether we're dealing with "tracing tensors".


[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][NT] Support a NJT constructed in a graph and returned from the graph"


We run into an error when we try to construct a NT in the graph and return it. There's two things fixed here:

1. some symint handling utils don't work with NestedInt (this is fixed by the changes in symbolic_shapes.py. error before the fixes: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py)
2. support tensor_unflatten for non-symbolic tensors. In this case, we haven't run mark_dynamic on the values or offsets, so torch.compile is treating them as static shape tensors. Although this is not ideal (we're almost certainly going to have to recompile on the next batch which will probably have different offsets / different total numel), it shouldn't be an error. In this case, this is relevant for some logic in tensor_unflatten: if we're dealing with a functional tensor or fake tensor and we're trying to tensor_unflatten using "tracing tensors" as the offsets/values, then we know we should be able to reuse the "old" NestedInt from the outer size. Previously, the check for "whether this is a tracing tensor" was "whether there are any symbolic sizes"; but in the case where we haven't run `mark_dynamic`, we don't have any symbolic sizes. Instead, just check if we have a FunctionalTensor or FakeTensor as an indicator of whether we're dealing with "tracing tensors".


[ghstack-poisoned]

soulitzer commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -214,7 +214,8 @@ def is_concrete_int(a: Union[int, SymInt]) -> bool:
 
-    if isinstance(a.node.expr, sympy.core.numbers.Integer):
+    # note: NestedInt is a SymInt but does not have a.node.expr defined.
+    if isinstance(getattr(a.node, "expr", None), sympy.core.numbers.Integer):
```
Do you know where this is called?
Nested int technically "concrete" and in many(most?) cases nested_int is intended to be used where int is used. Could we want to return True instead?

davidberard98 replied:
@soulitzer - I switched this to return True for NestedInts and it also fixes this issue (although the other issue with evaluation of whether a tensor is a tracing tensor still exists after changing this)

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -3997,6 +3997,22 @@ def check_size(nt1, nt2, nt3, nt4):
+
+        ref = fn(values, offsets)
+        res = fn_c(values, offsets)
+        self.assertEqual(ref, res)
```
It feels like this test should still pass without the changes below because our outer sizes are not symbolic in this case? 
e.g. when we create the nested int in the ctor during tracing, we'd go through get_nested_int and that would return a non-symbolic nested int.

davidberard98 replied:
> get_nested_int and that would return a non-symbolic nested int.

yes, but it gets a new nested int - so without this change, the outer_size is (4, j2, 8) and the returned tensor has (4, j3, 8)

davidberard98 replied:
The test wasn't actually failing before but I updated it to a version where it does fail (the offsets need to be created inside the graph)

soulitzer replied:
Hmmm interesting that it was working before - I'm playing around w/ something similar to the example, and ran into the error you described.
also, with the "whether a tensor is a tracing tensor" changes, it passes but still fails when grad is enabled

soulitzer replied:
```
from torch.nested._internal.nested_tensor import NestedTensor
import torch

def fn(values, offsets):
  nt =  torch.nested.nested_tensor_from_jagged(values, offsets)
  return nt.cos()

device="cuda"
values = torch.rand((10, 8), device=device)
offsets = torch.tensor([0, 1, 3, 6, 10], device=device)

out = torch.compile(fn, fullgraph=True, backend="aot_eager")(values, offsets)
```

soulitzer replied:
From https://github.com/pytorch/pytorch/pull/124624, to fix the case when the values input requires grad, 
```
diff --git a/torch/nested/_internal/nested_tensor.py b/torch/nested/_internal/nested_tensor.py
index 5494633b09b..017c000d99c 100644
--- a/torch/nested/_internal/nested_tensor.py
+++ b/torch/nested/_internal/nested_tensor.py
@@ -15,6 +15,9 @@ def get_tensor_symint(tensor, *, coeff=1):
     global _tensor_id_counter
     tensor_symint = _tensor_symint_registry.get(tensor)
     if tensor_symint is None:
+        if isinstance(tensor, torch._subclasses.functional_tensor.FunctionalTensor):
+            tensor = torch._from_functional_tensor(tensor.elem)
+            return get_tensor_symint(tensor, coeff=coeff)
         tensor_symint = torch._C._get_nested_int(_tensor_id_counter, coeff)
         _tensor_id_counter += 1
         _tensor_symint_registry[tensor] = tensor_symint
```
The main diffference between that PR and this one is that that one uses unbacked symints over non-symbolic nested int, which I think I prefer.

davidberard98 replied:
ah okay, I think I see how your suggestion works! and it does seem to fix this case as well

davidberard98 committed:
Update on "[NT] Support a NJT constructed in a graph and returned from the graph"


We run into an error when we try to construct a NT in the graph and return it. There's two things fixed here:

1. some symint handling utils don't work with NestedInt (this is fixed by the changes in symbolic_shapes.py. error before the fixes: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py)
2. support tensor_unflatten for non-symbolic tensors. In this case, we haven't run mark_dynamic on the values or offsets, so torch.compile is treating them as static shape tensors. Although this is not ideal (we're almost certainly going to have to recompile on the next batch which will probably have different offsets / different total numel), it shouldn't be an error. In this case, this is relevant for some logic in tensor_unflatten: if we're dealing with a functional tensor or fake tensor and we're trying to tensor_unflatten using "tracing tensors" as the offsets/values, then we know we should be able to reuse the "old" NestedInt from the outer size. Previously, the check for "whether this is a tracing tensor" was "whether there are any symbolic sizes"; but in the case where we haven't run `mark_dynamic`, we don't have any symbolic sizes. Instead, just check if we have a FunctionalTensor or FakeTensor as an indicator of whether we're dealing with "tracing tensors".


[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Support a NJT constructed in a graph and returned from the graph"


We run into an error when we try to construct a NT in the graph and return it. There's two things fixed here:

1. some symint handling utils don't work with NestedInt (this is fixed by the changes in symbolic_shapes.py. error before the fixes: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py)
2. support tensor_unflatten for non-symbolic tensors. In this case, we haven't run mark_dynamic on the values or offsets, so torch.compile is treating them as static shape tensors. Although this is not ideal (we're almost certainly going to have to recompile on the next batch which will probably have different offsets / different total numel), it shouldn't be an error. In this case, this is relevant for some logic in tensor_unflatten: if we're dealing with a functional tensor or fake tensor and we're trying to tensor_unflatten using "tracing tensors" as the offsets/values, then we know we should be able to reuse the "old" NestedInt from the outer size. Previously, the check for "whether this is a tracing tensor" was "whether there are any symbolic sizes"; but in the case where we haven't run `mark_dynamic`, we don't have any symbolic sizes. Instead, just check if we have a FunctionalTensor or FakeTensor as an indicator of whether we're dealing with "tracing tensors".


[ghstack-poisoned]

davidberard98 committed:
Update on "[NT] Support a NJT constructed in a graph and returned from the graph"


We run into an error when we try to construct a NT in the graph and return it. There's two things fixed here:

1. some symint handling utils don't work with NestedInt (this is fixed by the changes in symbolic_shapes.py. error before the fixes: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py)
2. support tensor_unflatten for non-symbolic tensors. In this case, we haven't run mark_dynamic on the values or offsets, so torch.compile is treating them as static shape tensors. Although this is not ideal (we're almost certainly going to have to recompile on the next batch which will probably have different offsets / different total numel), it shouldn't be an error. In this case, this is relevant for some logic in tensor_unflatten: if we're dealing with a functional tensor or fake tensor and we're trying to tensor_unflatten using "tracing tensors" as the offsets/values, then we know we should be able to reuse the "old" NestedInt from the outer size. Previously, the check for "whether this is a tracing tensor" was "whether there are any symbolic sizes"; but in the case where we haven't run `mark_dynamic`, we don't have any symbolic sizes. Instead, just check if we have a FunctionalTensor or FakeTensor as an indicator of whether we're dealing with "tracing tensors".


[ghstack-poisoned]

soulitzer commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -214,6 +214,10 @@ def is_concrete_int(a: Union[int, SymInt]) -> bool:
 
+    # note: NestedInt is a concerete int but does not have a.node.expr defined.
+    if a.node.is_nested_int():
+        return True
```
If I had a symbolic symnode holding nested int as hint, I think that also return True here, which we may not want.

soulitzer reviewed (commented):
Hmmmm do we still need this PR (since its specific to the case if we want to rely on non-symbolic nested int to exist on the NT during tracing). My other PR (which uses unbacked symint) didn't end up needing this change for the test to pass, at least for the cases I've tested.

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 committed:
Update on "[NT] Support NestedTensor in is_concrete_int"


We run into an error when we try to construct a NT in the graph and return it. One of the errors is with symint handling in is_concrete_int: the full error this PR is here: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py.

Note - a separate issue occurs if we don't call fn() before fn_c(), where the tensors aren't marked as dynamic; fix for this is in https://github.com/pytorch/pytorch/pull/124624

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58315861).

davidberard98 commented:
@pytorchbot rebase

davidberard98 committed:
Update on "[NT] Support NestedTensor in is_concrete_int"


We run into an error when we try to construct a NT in the graph and return it. One of the errors is with symint handling in is_concrete_int: the full error this PR is here: https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da. Repro: run the test_nestedtensor.py test after reverting the patch in symbolic_shapes.py.

Note - a separate issue occurs if we don't call fn() before fn_c(), where the tensors aren't marked as dynamic; fix for this is in https://github.com/pytorch/pytorch/pull/124624

Differential Revision: [D58315861](https://our.internmc.facebook.com/intern/diff/D58315861)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58315861).

davidberard98 commented:
@soulitzer do you have more progress on #124687? what do you think about landing this?

I think we're starting to see this issue in some of the experiments that @YuqingJ and I have run, so it would be good to have a patch. I think @jbschlosser also needed this to get rid of graph breaks with SDPA + NT in his test here: https://github.com/pytorch/pytorch/issues/126472#issuecomment-2118257497?

davidberard98 commented:
@pytorchbot rebase

davidberard98 committed:
Update

[ghstack-poisoned]
